File: warrior/src/main/java/com/dumbdogdiner/warrior/managers/KitManager.java
package com.dumbdogdiner.warrior.managers;

import com.dumbdogdiner.warrior.Warrior;
import com.dumbdogdiner.warrior.api.kit.BaseKit;
import com.dumbdogdiner.warrior.api.kit.kits.ArcherKit;
import com.dumbdogdiner.warrior.api.kit.kits.CustomKit;
import com.dumbdogdiner.warrior.api.kit.kits.TankKit;
import com.dumbdogdiner.warrior.api.kit.kits.WarriorKit;
import com.dumbdogdiner.warrior.api.managers.WarriorKitManager;
import com.dumbdogdiner.warrior.models.CustomKitModel;
import com.dumbdogdiner.warrior.api.translation.Placeholders;
import com.dumbdogdiner.warrior.api.util.JSONUtil;
import com.dumbdogdiner.warrior.utils.DefaultMessages;
import com.google.gson.Gson;
import com.google.gson.stream.JsonReader;
import org.bukkit.Material;

import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;


public class KitManager implements WarriorKitManager {

    private final ArrayList<BaseKit> kits = new ArrayList<>();

    public void registerKits() {
        kits.add(new WarriorKit("Warrior", 0, "warrior.kit.warrior", Material.IRON_SWORD));
        kits.add(new ArcherKit("Archer", 0, "warrior.kit.archer", Material.BOW));
        kits.add(new TankKit("Tank", 0, "warrior.kit.tank", Material.IRON_CHESTPLATE));

        loadFiles();
    }

    /**
     * Load kits from plugin configuration.
     */
    private void loadFiles() {
        File dataFolder = new File(JSONUtil.KIT_DATA_PATH);
        File[] files = dataFolder.listFiles();

        if(files == null) return;

        int i = 0;
        for(File f : files) {
            addKit(f);
            i++;
        }

        int finalI = i;
        String msg = Placeholders.applyPlaceholders(DefaultMessages.LOADED_OBJECT, new HashMap<>() {
            {
                put("AMOUNT", Integer.toString(finalI));
                put("TYPE", "kit");
            }
        });
        Warrior.getPluginLogger().info(msg);
    }

    public void addKit(File f) {
        if(!f.canRead()) return;
        try(JsonReader reader = new JsonReader(new FileReader(f))) {
            CustomKitModel model = new Gson().fromJson(reader, CustomKitModel.class);
            kits.add(new CustomKit(model));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void addKit(CustomKit kit) {
        kits.add(kit);
    }

    public BaseKit get(String n) {
        for(BaseKit k : kits) {
            if(k.getName().equalsIgnoreCase(n)) return k;
        }
        return kits.get(0);
    }

    public List<BaseKit> getKits() {
        return kits;
    }
}

File: warrior/src/main/kotlin/com/dumbdogdiner/warrior/managers/KitManager.kt
package com.dumbdogdiner.warrior.managers

import com.dumbdogdiner.warrior.api.managers.WarriorKitManager
import com.dumbdogdiner.warrior.api.kit.BaseKit
import com.dumbdogdiner.warrior.api.kit.kits.WarriorKit
import com.dumbdogdiner.warrior.api.kit.kits.ArcherKit
import com.dumbdogdiner.warrior.api.kit.kits.TankKit
import com.dumbdogdiner.warrior.api.translation.Placeholders
import com.dumbdogdiner.warrior.util.DefaultMessages
import java.util.HashMap
import com.dumbdogdiner.warrior.Warrior
import java.io.FileReader
import com.dumbdogdiner.warrior.api.kit.kits.CustomKit
import com.dumbdogdiner.warrior.api.util.json.models.CustomKitModel
import com.dumbdogdiner.warrior.util.JSONUtil
import com.google.gson.Gson
import com.google.gson.stream.JsonReader
import org.bukkit.Material
import java.io.File
import java.io.IOException
import java.util.ArrayList

class KitManager : WarriorKitManager {
    private val kits = ArrayList<BaseKit>()

    fun registerKits() {
        kits.add(WarriorKit("Warrior", 0, "warrior.kit.warrior", Material.IRON_SWORD))
        kits.add(ArcherKit("Archer", 0, "warrior.kit.archer", Material.BOW))
        kits.add(TankKit("Tank", 0, "warrior.kit.tank", Material.IRON_CHESTPLATE))
        loadFiles()
    }

    /**
     * Load kits from plugin configuration.
     */
    private fun loadFiles() {
        val dataFolder = File(JSONUtil.KIT_DATA_PATH)
        val files = dataFolder.listFiles() ?: return
        var i = 0

        for (f in files) {
            addKit(f)
            i++
        }

        val msg = Placeholders.applyPlaceholders(DefaultMessages.LOADED_OBJECT, object : HashMap<String?, String?>() {
            init {
                put("AMOUNT", i.toString())
                put("TYPE", "kit")
            }
        })

        Warrior.pluginLogger.info(msg)
    }

    override fun addKit(f: File) {
        if (!f.canRead()) return

        try {
            JsonReader(FileReader(f)).use { reader ->
                val model = Gson().fromJson<CustomKitModel>(reader, CustomKitModel::class.java)
                kits.add(CustomKit(model))
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    override fun addKit(kit: CustomKit) {
        kits.add(kit)
    }

    override fun get(n: String): BaseKit {
        for (k in kits) {
            if (k.name.equals(n, ignoreCase = true)) return k
        }
        return kits[0]
    }

    override fun getKits(): List<BaseKit> {
        return kits
    }
}